### PR TITLE
crimson/osd: do not pass oop as a const ref and cleanups

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/osd_operations/client_request.cc [hunk 1]
```diff
@@ -116,70 +116,81 @@ seastar::future<> ClientRequest::process_pg_op(
...
-        return process_op(pgref);
-      }));
+  return with_blocking_future(handle.enter(pp(*pg).recover_missing)).then([&] {
+    return do_recover_missing(pg);
+  }).then([&] {
+    return with_blocking_future(handle.enter(pp(*pg).get_obc));
+  }).then([this, &pg]() -> PG::load_obc_ertr::future<> {
+    op_info.set_from_op(&*m, *pg->get_osdmap());
+    return pg->with_locked_obc(m, op_info, this, [this, &pg](auto obc) {
+      op_info.set_from_op(&*m, *pg->get_osdmap());
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/client_request.cc:128]: Duplicated call to `set_from_op()`.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 2]
```diff
@@ -116,70 +116,81 @@ seastar::future<> ClientRequest::process_pg_op(
...
+    return std::move(fut);
+  }
+}
+
+seastar::future<>
+ClientRequest::do_process(Ref<PG>& pg, crimson::osd::ObjectContextRef obc)
+{
+  using do_ops_return_t =
+    crimson::errorator<crimson::ct_error::eagain>::future<Ref<MOSDOpReply>>;
+  return [=]() -> do_ops_return_t {
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/client_request.cc]: Is there a good reason for `=`? Maybe the dependencies are limited enough to catch them explicitly, one-by-one?

File: src/crimson/osd/osd_operations/client_request.cc [hunk 3]
```diff
@@ -162,30 +162,22 @@ seastar::future<> ClientRequest::do_recover_missing(Ref<PG>& pg)
 seastar::future<>
 ClientRequest::do_process(Ref<PG>& pg, crimson::osd::ObjectContextRef obc)
 {
-  using do_ops_return_t =
-    crimson::errorator<crimson::ct_error::eagain>::future<Ref<MOSDOpReply>>;
-  return [=]() -> do_ops_return_t {
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/client_request.cc]: OK, refactored.

File: src/crimson/osd/osd_operations/client_request.cc [hunk 4]
```diff
@@ -162,30 +162,22 @@ seastar::future<> ClientRequest::do_recover_missing(Ref<PG>& pg)
...
+    if (is_misdirected(*pg)) {
+      logger().trace("process_op: dropping misdirected op");
       return seastar::now();
+    } else if (const hobject_t& hoid = m->get_hobj();
+               !pg->get_peering_state().can_serve_replica_read(hoid)) {
+      auto reply = make_message<MOSDOpReply>(
+	m.get(), -EAGAIN, pg->get_osdmap_epoch(),
+	m->get_flags() & (CEPH_OSD_FLAG_ACK|CEPH_OSD_FLAG_ONDISK),
+	!m->has_flag(CEPH_OSD_FLAG_RETURNVEC));
+      return conn->send(std::move(reply));
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/client_request.cc:176]: Much cleaner that way :+1:.

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/osd/osd_operations/client_request.cc (hunk 1 above)
Comment (tchaikov) [reply src/crimson/osd/osd_operations/client_request.cc]: thanks. removed.

File: src/crimson/osd/osd_operations/client_request.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/osd/osd_operations/client_request.cc]: no. should have captured `pg` by reference. fixed and repushed.

**Review by rzarzynski — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
rebased against master.

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
changelog

- piggy-back some changes to address FTBFS with clang. these changes should have been included in #39359. but i missed them somehow.

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
[ RUN      ] mempool.check_shard_select
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/test_mempool.cc:430: Failure
Expected: (VarX) < (200), actual: 221.5 vs 200
[  FAILED  ] mempool.check_shard_select (37 ms)
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Review by xxhdx1985126 — APPROVED** [review]

**Comment by tchaikov** [comment]
added a commit of "crimson/os: do not capture unused variables".

**Pull request merged**